src/outputs/rcp.c

```diff
@@ -0,0 +1,1484 @@
+      case RCP_STATE_VOL_GET:
+      case RCP_STATE_VOL_SET:
+      case RCP_STATE_STOPPING ... RCP_STATE_SHUTDOWN_SERVER_DISCONNECT:  // as per @frankusb, move advise that remote/we're not interested
+	state = OUTPUT_STATE_CONNECTED;
+	break;
+
+      case RCP_STATE_STREAMING:
+	state = OUTPUT_STATE_STREAMING;
+	break;
+
+      case RCP_STATE_STOPPED ... RCP_STATE_DISCONNECTED:
+	state = OUTPUT_STATE_STOPPED;
+	break;
+
+      case RCP_STATE_FAILED:
+      default:
+	state = OUTPUT_STATE_FAILED;
+    }
+
+  DPRINTF(E_DBG, L_RCP, "state -> %d\n", state);
+  outputs_cb(s->callback_id, s->device->id, state);
+  s->callback_id = -1;
+
+  if (state == OUTPUT_STATE_STOPPED || state == OUTPUT_STATE_FAILED)
+    rcp_session_cleanup(s);
+}
+
+
+/* ------------------ INTERFACE FUNCTIONS CALLED BY OUTPUTS.C --------------- */
+
+static int
+rcp_device_start(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "device start %ld '%s'\n", device->id, device->name);
+
+  s = rcp_session_make(device, callback_id);
+  if (!s)
+    return -1;
+
+  rcp_status(s);
+
+  return 1;
+}
+
+static int
+rcp_device_stop(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+
+  DPRINTF(E_DBG, L_RCP, "device stop %ld '%s'\n", device->id, device->name);
+
+  /* force these devices as deselected (auto state saves in db later) since 
+   * these need use to select (and cause the device probe to start connection to
+   * remote side
+   */
+  device->selected = 0;
+
+  s->callback_id = callback_id;
+  // tear this session down, incl free'ing it
+  rcp_session_shutdown(s, RCP_STATE_STOPPED);
+
+  return 1;
+}
+
+static int
+rcp_device_flush(struct output_device *device, int callback_id)
+{
+  DPRINTF(E_DBG, L_RCP, "device flush %ld '%s'\n", device->id, device->name);
+
+  // needed?
+  struct rcp_session *s = device->session;
+
+  s->callback_id = callback_id;
+  s->state = OUTPUT_STATE_STOPPED;
+
+  rcp_status(s);
+
+  return 1;
+}
+
+static int
+rcp_device_probe(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "device probe %ld '%s'\n", device->id, device->name);
+
+  s = rcp_session_make(device, callback_id);
+  if (!s)
+    return -1;
+
+  s->callback_id = callback_id;
+  s->state = OUTPUT_STATE_STOPPED;
+
+  rcp_status(s);
+
+  return 1;
+}
+
+
+static int
+rcp_device_volume_set(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+  char cmd[3];
+  int ret;
+
+  DPRINTF(E_DBG, L_RCP, "device volume set %ld '%s' state=%d vol=%d\n", device->id, device->name, s->state, device->volume);
+
+  if (!s || (s && s->state != RCP_STATE_STREAMING))
+    return 0;
+
+  s->callback_id = callback_id;
+
+
+  ret = snprintf(cmd, sizeof(cmd), "%d", device->volume);
+  if (ret < 0) {
+    return 0;
+  }
+
+  rcp_send(s, RCP_STATE_VOL_SET, cmd);
+
+  rcp_status(s);
```

Comment:
No ok from device?

Reply:
moving these into `rcp_status_verify()` which is only called when a full response recvied